nuxeo-features/nuxeo-platform-tag-service/nuxeo-platform-tag-core/src/main/java/org/nuxeo/ecm/platform/tag/FacetedTagService.java

```diff
@@ -0,0 +1,153 @@
+/*
+ * (C) Copyright 2017 Nuxeo (http://nuxeo.com/) and others.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Contributors:
+ *     Funsho David
+ */
+
+package org.nuxeo.ecm.platform.tag;
+
+import org.nuxeo.ecm.core.api.CoreSession;
+import org.nuxeo.ecm.core.api.DocumentModel;
+import org.nuxeo.ecm.core.api.IdRef;
+import org.nuxeo.ecm.core.versioning.VersioningService;
+
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static org.nuxeo.ecm.core.api.CoreSession.ALLOW_VERSION_WRITE;
+import static org.nuxeo.ecm.core.query.sql.NXQL.ECM_UUID;
+import static org.nuxeo.ecm.platform.tag.TagConstants.TAG_FACET;
+import static org.nuxeo.ecm.platform.tag.TagConstants.TAG_LIST;
+
+/**
+ * Implementation of the tag service based on facet
+ *
+ * @since 9.3
+ */
+public class FacetedTagService extends AbstractTagService {
+
+    @Override
+    public void doTag(CoreSession session, String docId, String label) {
+        DocumentModel docModel = session.getDocument(new IdRef(docId));
+        if (!docModel.isProxy()) {
```

Comment:
There should probably be an `else` throwing an exception saying that tagging proxies is not supported, instead of failing silently. And add a test for that.